[Zheng Tao Jun] iP

TaoJun99 opened the pull request:
# Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn 
* ~~FAST~~ *SUPER FAST* to use

All you need to do is,
1. download iP.jar file from [here](https://github.com/TaoJun99/ip/releases)
2. double-click it
3. add your tasks
4. let it manage your tasks for you :smiley:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public static void main(String[] args) {
        new Duke("data/duke.txt").run();
 }
```

Commit:
Add Gradle support

Commit:
Level-1

Commit:
Level-2

Commit:
Level-3

Commit:
Level-4

Commit:
A-TextUiTesting

Commit:
Level-5

Commit:
Level-6

Commit:
A-Enums

Commit:
Level-7

Commit:
Dates and Times

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Dates and Times

Commit:
Fix file's date and time loading issues

Commit:
Update input.txt and EXPECTED.txt

Commit:
Extract Ui, Storage, Parser, TaskList classes

Commit:
Divide classes into packages

Commit:
Add JUnit tests

Commit:
Package app as JAR file

Commit:
Add JavaDoc comments

Commit:
Tweak the code to comply with coding standard

Commit:
Add command "find" to search task

Commit:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak the code to comply with coding standard

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

Commit:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add command "find" to search task

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Ui.java

Commit:
Merge branch 'add-gradle-support' of https://github.com/TaoJun99/ip into A-Gradle

Commit:
Automate project builds using Gradle

Commit:
Use CheckStyle

Commit:
Add GUI to Duke

limdanqi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    public Command parseCommand(String input) throws InvalidTaskException {
+        String parsed = input.split(" ", 2)[0];
+        boolean validCommand = false;
```
```suggestion
boolean isValidCommand = false;
```

limdanqi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,105 @@
+     * @throws NoDescriptionException if task number not included in user input
+     */
+    public int getTaskNo(String input) throws NoDescriptionException {
+        String[] parsed = input.split(" ", 2);
```
```suggestion
String[] parsedStrings = input.split(" ", 2);
```

limdanqi commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public class Deadline extends Task {
+    protected LocalDateTime by;
+    private final DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("E, dd MMM yyyy HHmm");
```
```suggestion
private final DateTimeFormatter OUTPUT_FORMATTER = DateTimeFormatter.ofPattern("E, dd MMM yyyy HHmm");
```

limdanqi commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,61 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
```
A possible suggestion would be to add header comments for the tester classes too.

limdanqi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,105 @@
+     * @throws NoDescriptionException if description not included in user input
+     * @throws WrongDescriptionException if conjunction not included in user input
+     */
+
```
The line break here can be deleted.

limdanqi reviewed (commented):
Overall, I found your code to be pretty impressive. The Duke bot is able to function well and the code was concise and clean. I also like the organization of the code itself as it makes reading it much easier. In general, there were only a few minor coding violations. Keep it up! :D

Commit:
Improve GUI

idgrr commented on src/main/java/duke/gui/MainWindow.java:
```diff
@@ -0,0 +1,66 @@
+        );
+        userInput.clear();
+        if (input.equals("bye")) {
+            new Timer().schedule(new TimerTask() {
```
Nice idea to have a short-timer before closing

idgrr commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,111 @@
+package duke;
+
+import duke.gui.Main;
```
Nice organisation of files!

idgrr commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,107 @@
+     * @return done message
+     */
+    public String showDoneMessage(Task doneTask) {
+        return "Nice! I've marked this task as done:\n  " + doneTask;
```
May i know why on some show*** methods you use String.format and others u use String addition?

idgrr commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        if (!validCommand) {
+            throw new InvalidTaskException("Invalid command! Please enter the following commands only:\n"
```
I like that possible solutions were also provided in the error message. Nice!

idgrr reviewed (commented):
Overall great job! The code was clean and easy to follow.

Commit:
Add assert feature

Assert feature is not used.

Assert feature can help to document important assumptions that should hold at various points in the code.

Let's,
* enable assert feature in gradle
* add assert feature to Duke, Storage and Ui classes

TaoJun99 committed:
Merge pull request #1 from TaoJun99/branch-A-Assertions

Add assert feature

Commit:
Improve code quality

Current code has some code quality issues

Code of better quality is more readable and easier to understand.

Let's,
* organise the code
* minimise duplicated code

Commit:
Add C-Update extension

User is unable to edit tasks

Adding this extension introduces a new command so that users are able to edit tasks in the task list in case of wrongly written tasks or changing deadlines or event times.

TaoJun99 committed:
Merge pull request #2 from TaoJun99/branch-A-CodeQuality

Improve code quality

Commit:
Merge branch 'master' into branch-C-Update

* master:
  Improve code quality

# Conflicts:
#	data/duke.txt
#	src/main/java/duke/Parser.java

TaoJun99 committed:
Merge pull request #3 from TaoJun99/branch-C-Update

Branch c update

Commit:
Improve the GUI to make it more polished

Commit:
Fixed some OOP issues

Commit:
Add a User Guide

Commit:
Fix some minor code quality issues

Commit:
Fix 'Java has been detached' error

Commit:
Update User Guide
